algorithms/src/snark/marlin/marlin.rs

```diff
@@ -191,7 +218,9 @@ impl<E: PairingEngine, FS: AlgebraicSponge<E::Fq, 2>, MM: MarlinMode> MarlinSNAR
     fn absorb_with_msg(commitments: &[Commitment<E>], msg: &prover::ThirdMessage<E::Fr>, sponge: &mut FS) {
```

Comment:
note on the `absorb` family of methods (and `PoseidonSponge::absorb_native_field_elements`): I can see that they could all take their slice params as iterators, which could lead to a reduction in allocations

Reply:
Good idea, but outside of scope of this (monster) PR, let's await future benchmarks/perf results before tackling this